File: python-probes/src/soma_probes/flax/v1/modules/layer.py
from jax import Array
from flax import nnx

from soma_probes.flax.v1.modules.pwff import (
    PositionWiseFeedForward,
    PositionWiseFeedForwardConfig,
)
from soma_probes.flax.v1.modules.encoder import EncoderConfig
from soma_probes.flax.v1.modules.attention import MultiHeadAttention


class Layer(nnx.Module):
    def __init__(self, config: EncoderConfig, rngs: nnx.Rngs):
        self.norm_1 = nnx.LayerNorm(
            num_features=config.embedding_dim,
            rngs=rngs,
            epsilon=1e-5,
            use_fast_variance=False,
        )
        self.attention = MultiHeadAttention(
            num_heads=config.num_heads,
            in_features=config.embedding_dim,
            dropout_rate=config.dropout_rate,
            decode=False,
            max_wavelength=config.max_wavelength,
            rngs=rngs,
        )
        self.norm_2 = nnx.LayerNorm(
            num_features=config.embedding_dim,
            rngs=rngs,
            use_fast_variance=False,
            epsilon=1e-5,
        )
        self.pwff = PositionWiseFeedForward(
            PositionWiseFeedForwardConfig(dropout_rate=config.dropout_rate), rngs
        )
        self.dropout = nnx.Dropout(config.dropout_rate, rngs=rngs)

    def __call__(
        self,
        representations: Array,
        positions: Array,
    ):
        x = representations
        residual_path = self.norm_1(x)
        residual_path = self.attention(residual_path, positions)
        residual_path = self.dropout(residual_path)
        x = x + residual_path
        residual_path = self.norm_2(x)
        residual_path = self.pwff(residual_path)
        residual_path = self.dropout(residual_path)
        x = x + residual_path
        return x

File: models/src/v1/modules/layer.rs
// Copyright (c) Soma Contributors
// SPDX-License-Identifier: Apache-2.0

use super::{
    encoder::EncoderConfig,
    pwff::{PositionWiseFeedForward, PositionWiseFeedForwardConfig},
};
use crate::v1::modules::attention::{MhaInput, MultiHeadAttention, MultiHeadAttentionConfig};
use burn::{
    module::Module,
    nn::{LayerNorm, LayerNormConfig},
    tensor::{Int, Tensor, backend::Backend},
};

#[derive(Module, Debug)]
pub struct Layer<B: Backend> {
    norm_1: LayerNorm<B>,
    attention: MultiHeadAttention<B>,
    norm_2: LayerNorm<B>,
    pwff: PositionWiseFeedForward<B>,
}

impl<B: Backend> Layer<B> {
    pub(crate) fn new(config: &EncoderConfig, device: &B::Device) -> Self {
        Layer {
            norm_1: LayerNormConfig::new(config.embedding_dim).init(device),
            attention: MultiHeadAttentionConfig::new()
                .with_num_features(config.embedding_dim)
                .with_num_heads(config.num_heads)
                .with_initializer(config.initializer.clone())
                .init(device),
            norm_2: LayerNormConfig::new(config.embedding_dim).init(device),
            pwff: PositionWiseFeedForwardConfig::new()
                .with_embedding_dim(config.embedding_dim)
                .with_hidden_dim(config.pwff_hidden_dim)
                .with_initializer(config.initializer.clone())
                .init(device),
        }
    }
    pub(crate) fn forward(
        &self,
        context: Tensor<B, 3>,
        positions: Tensor<B, 2, Int>,
    ) -> Tensor<B, 3> {
        let x = context;
        let residual_path = self.norm_1.forward(x.clone());
        let input_mha = MhaInput::new(residual_path, positions);
        let residual_path = self.attention.forward(input_mha);
        let x = x + residual_path;
        let residual_path = self.norm_2.forward(x.clone());
        let residual_path = self.pwff.forward(residual_path);
        x + residual_path
    }
}
